core-rust/state-manager/src/staging/cache.rs

```diff
@@ -420,17 +431,108 @@ impl<K, N> AccuTreeDiff<K, N> {
 }
 
+#[derive(Clone)]
+pub enum ImmutablePartitionUpdates {
+    Delta {
+        substate_updates: ImmutableHashMap<DbSortKey, DatabaseUpdate>,
+    },
+    Batch(BatchImmutablePartitionUpdates),
+}
+
+impl Default for ImmutablePartitionUpdates {
+    fn default() -> Self {
+        Self::Delta {
+            substate_updates: ImmutableHashMap::new(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum BatchImmutablePartitionUpdates {
+    Reset {
+        new_substate_values: ImmutableHashMap<DbSortKey, DbSubstateValue>,
+    },
+}
+
+impl ImmutablePartitionUpdates {
+    pub fn accumulate(&mut self, db_partition_updates: &PartitionDatabaseUpdates) {
+        match self {
+            ImmutablePartitionUpdates::Delta { substate_updates } => match db_partition_updates {
+                PartitionDatabaseUpdates::Delta {
+                    substate_updates: db_substate_updates,
+                } => {
+                    substate_updates.extend(
+                        db_substate_updates
+                            .iter()
+                            .map(|(key, value)| (key.clone(), value.clone())),
+                    );
+                }
+                PartitionDatabaseUpdates::Reset {
+                    new_substate_values: db_new_substate_values,
+                } => {
+                    *self =
+                        ImmutablePartitionUpdates::Batch(BatchImmutablePartitionUpdates::Reset {
+                            new_substate_values: db_new_substate_values
+                                .iter()
+                                .map(|(key, value)| (key.clone(), value.clone()))
+                                .collect(),
+                        })
+                }
+            },
+            ImmutablePartitionUpdates::Batch(batch) => {
+                batch.accumulate(db_partition_updates);
+            }
+        }
+    }
+}
+
+impl BatchImmutablePartitionUpdates {
+    pub fn accumulate(&mut self, db_partition_updates: &PartitionDatabaseUpdates) {
+        match self {
+            BatchImmutablePartitionUpdates::Reset {
+                new_substate_values,
+            } => match db_partition_updates {
+                PartitionDatabaseUpdates::Delta {
+                    substate_updates: db_substate_updates,
+                } => {
+                    for (db_sort_key, database_update) in db_substate_updates {
+                        match database_update {
+                            DatabaseUpdate::Set(value) => {
+                                new_substate_values.insert(db_sort_key.clone(), value.clone());
+                            }
+                            DatabaseUpdate::Delete => {
+                                let existed = new_substate_values.remove(db_sort_key).is_some();
+                                if !existed {
+                                    panic!("broken invariant: deleting non-existent substate");
```

Comment:
_(discussed on a call, but for the record: we should double check that the same invariant always holds in the engine)_